kvserver: decline rebalance snapshots on receivers with poor LSM

aayushshah15 opened the pull request:
First commit from https://github.com/cockroachdb/cockroach/pull/77246.

This patch introduces two new cluster settings:
```
kv.snapshot_decline.read_amp_threshold

server.declined_snapshot_timeout
```

With this commit, stores with a read amplification level higher than
`kv.snapshot_decline.read_amp_threshold` will decline all `REBALANCE`
snapshots. Upon receiving a `DECLINED` response, the senders of these snapshots
will consider these receivers `throttled` for `server.declined_snapshot_timeout`.

This means that stores with poor LSM health will not be considered as valid
candidates for replica rebalancing.

Fixes #73714
Related to #62168

Release justification: This patch adds a tunable guardrail that could prevent 
or mitigate cluster instability.

Release note: None

\cc. @cockroachdb/storage

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73720)
<!-- Reviewable:end -->

aayushshah15 commented:
@sumeerbhola, I noticed that we already have somewhat [analogous cluster settings for admission control](https://github.com/cockroachdb/cockroach/blob/465c8eaaacd8b8219e89795577fc21f60a96de3d/pkg/util/admission/granter.go#L38-L50). However, I am not sure that we want to tie the admission control behavior and this rebalancing behavior onto the same set of settings. Do you have any thoughts on that?

sumeerbhola reviewed (commented):
I have a few questions on the approach:
- How much work has been done on the source of the snapshot before it receives a decline. Looks like the Pebble snapshot has been taken, yes? Has much data been read using the iterator?
- If every node in the cluster is considered unhealthy and a node(s) is being shutdown, does this mean we could have a loss of quorum? That is, would it be better for the allocator to decide to rebalance to the least unhealthy node, instead of the target unilaterally deciding to reject.

Regarding the `kv.snapshot_decline.read_amp_threshold` of 1000, at a high-level I think admission control in these cases is a blunt hammer due to the very narrow scope of actions it can take (queue reordering, and delaying). So higher level logic that has more context should kick in before admission control thresholds are reached. That would apply to rebalancing and to https://github.com/cockroachdb/cockroach/issues/57247.

One possibility would be to make this 1000 read-amp cluster setting be a stop-gap until the allocator chooses the least unhealthy node if all are unhealthy (say with an unhealthy read amp threshold of 15), since I can see why lowering the current setting with this unilateral reject behavior could be risky.

> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-:-Mqog9TL2Jp9Xs1IPdlc:b-4sfnt3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
I think this is going to need a version gate, since older nodes won't understand this response and error out.

aayushshah15 commented:
Thanks for taking a look!

>How much work has been done on the source of the snapshot before it receives a decline. Looks like the Pebble snapshot has been taken, yes? Has much data been read using the iterator?

We only start [reading](https://github.com/cockroachdb/cockroach/blob/98116f3d8f069d5d3e7dd575741a1e507e1169d6/pkg/kv/kvserver/store_snapshot.go#L1005-L1009) once the receiver store accepts it by acquiring a "reservation" for this snapshot (adhering to a concurrency limit on the receiver side): https://github.com/cockroachdb/cockroach/blob/98116f3d8f069d5d3e7dd575741a1e507e1169d6/pkg/kv/kvserver/store_snapshot.go#L943-L954

>If every node in the cluster is considered unhealthy and a node(s) is being shutdown, does this mean we could have a loss of quorum? That is, would it be better for the allocator to decide to rebalance to the least unhealthy node, instead of the target unilaterally deciding to reject.

We wouldn't have a loss of quorum, no. If every node in the system was unhealthy, we would effectively stall all replica rebalancing in the cluster until at least one node isn't.

>One possibility would be to make this 1000 read-amp cluster setting be a stop-gap until the allocator chooses the least unhealthy node if all are unhealthy (say with an unhealthy read amp threshold of 15), since I can see why lowering the current setting with this unilateral reject behavior could be risky.

This approach would require every store to advertise its LSM health (probably via gossip) to every other store. The ambiguous bit here is about where this LSM health attribute would rank relative to the [other signals](https://github.com/cockroachdb/cockroach/blob/6c829c25058a5bbb5b927cc13994b318df076cfe/pkg/kv/kvserver/allocator_scorer.go#L302-L312) that the allocator uses to pick and rank candidate stores for replica rebalancing. We could explore that approach, but it wouldn't be backportable.

aayushshah15 commented:
>I think this is going to need a version gate, since older nodes won't understand this response and error out.

I did consider that, but I dont think we need it. Older stores that do not understand this response will [still throttle the target](https://github.com/cockroachdb/cockroach/blob/541adf9e2a0a68d27a5fb7cfcb0003555ca1317c/pkg/kv/kvserver/store_snapshot.go#L928-L933) store, albeit only for 5 seconds. So I dont think the version gate gets us anything. Furthermore, we won't be able to backport this patch if we had a version gate.

erikgrinaker reviewed (approved):
> If every node in the cluster is considered unhealthy and a node(s) is being shutdown, does this mean we could have a loss of quorum? That is, would it be better for the allocator to decide to rebalance to the least unhealthy node, instead of the target unilaterally deciding to reject.
> 
> We wouldn't have a loss of quorum, no. If every node in the system was unhealthy, we would effectively stall all replica rebalancing in the cluster until at least one node isn't.

Yeah, I think the point to note here is that this only applies to snapshots with priority `REBALANCE`, not `RECOVERY`. So if we lose a node while all nodes are unhealthy, we'll still upreplicate the range.

>I did consider that, but I dont think we need it. Older stores that do not understand this response will still throttle the target store, albeit only for 5 seconds.

Yeah -- mixed-version clusters will get log messages about invalid responses though, but maybe that's fine.

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-:-MqpKcd07YkQy2QxEp9u:bgk9p6k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @nvanbenschoten)

___
*[pkg/kv/kvserver/store_pool.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-MqpMOPM-B3iPZbDXRFV:-MqpMOPM-B3iPZbDXRFW:bqji1y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/98116f3d8f069d5d3e7dd575741a1e507e1169d6/pkg/kv/kvserver/store_pool.go#L59)):*
> ```Go
> var DeclinedSnapshotTimeout = settings.RegisterDurationSetting(
> 	"server.declined_snapshot_timeout",
> 	"the amount of time to consider the store throttled for up-replication after a declined rebalance snapshot",
> ```

This only applies to rebalancing, not up-replication of under-replicated ranges, right? I wouldn't use the term "up-replication" here then.
___
*[pkg/kv/kvserver/store_snapshot.go, line 712 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-MqpLWCoBbNXQhCAsaGq:-MqpLWCoBbNXQhCAsaGr:b-sjglg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/98116f3d8f069d5d3e7dd575741a1e507e1169d6/pkg/kv/kvserver/store_snapshot.go#L712)):*
> ```Go
> 	return stream.Send(&SnapshotResponse{
> 		Status:  SnapshotResponse_DECLINED,
> 		Message: "snapshot decline due to poor LSM health",
> ```

nit: "declined"


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
thanks for taking a look, Erik

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-:-MqpQs0u2ZvDoCxdCum8:b7suims)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @nvanbenschoten)

___
*[pkg/kv/kvserver/store_pool.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-MqpMOPM-B3iPZbDXRFV:-MqpQqwn9zL1DruaaN7o:b-rnrapz) ([raw file](https://github.com/cockroachdb/cockroach/blob/98116f3d8f069d5d3e7dd575741a1e507e1169d6/pkg/kv/kvserver/store_pool.go#L59)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

This only applies to rebalancing, not up-replication of under-replicated ranges, right? I wouldn't use the term "up-replication" here then.
</blockquote></details>

done. 
___
*[pkg/kv/kvserver/store_snapshot.go, line 712 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-MqpLWCoBbNXQhCAsaGq:-MqpQyZP4Vxm7gJqs8ng:b-rnrapz) ([raw file](https://github.com/cockroachdb/cockroach/blob/98116f3d8f069d5d3e7dd575741a1e507e1169d6/pkg/kv/kvserver/store_snapshot.go#L712)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: "declined"
</blockquote></details>

done. 


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (for this backportable approach, and answers to my previous questions)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-:-MquSFiR49BjiM1jwzlv:b-1tlyue)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

irfansharif commented:
Bump, we saw  #73714 in a recent escalation. Are we planning to get this in? + backport?

aayushshah15 commented:
> Are we planning to get this in? + backport?

@nvanbenschoten pointed out that this approach is less than desirable since we do this snapshot handshake _after_ creating uninitialized learner replicas. This means that if the snapshot is rejected, those learners have to be rolled back, which is [best effort](https://github.com/cockroachdb/cockroach/blob/b88a2b797e7906d5465eb58681d5f61170af2500/pkg/kv/kvserver/replica_command.go#L1767). So the concern here was that we're risking potentially leaking an uninitialized replica, which has to later get cleared up by the replicaGC queue. 

As an alternative, we've been considering https://github.com/cockroachdb/cockroach/issues/74254.

In that escalation, was it just that we suspected #73714 to be making things worse or did we notice something more direct?

irfansharif commented:
> In that escalation, was it just that we suspected #73714 to be making things worse or did we notice something more direct?

It was not the direct cause of the badness, nor did it terribly worsen the problem. We observed a few nodes with high read-amp/misshapen LSMs; we had to drain their leases away to give it breathing room for compactions/reduce foreground impact, + compact the node offline. In the midst of all this we just observed that occasionally replicas were being balanced onto malperforming nodes.

shralex commented:
@aayushshah15 I think these are complementary. Gossiping more info about node health will help avoid trying to rebalance to unhealthy nodes. But we can't expect gossip to be instantaneous or perfect (even if we had centralized monitoring, there would be some delay) so your PR here is valuable. I suggest to complete this work and merge it.

nvb reviewed (commented):
I agree that we would benefit from both approaches. However, as Aayush pointed out, there is risk in landing this first and pushing a release where this is the only change we have. We have seen in recent cluster outages that an excess of uninitialized replicas can be destabilizing. The concern with this change on its own is that a rejected snapshot leaks an uninitialized replica (a small amount of on-disk and in-mem metadata, but non-negligible cost in https://github.com/cockroachlabs/support/issues/1340) and we do nothing to eventually stem the tide of incoming snapshot requests. So we could spin on rebalance attempts indefinitely and amass a very large number of uninitialized replicas.

That said, there are mitigating factors. For one, we have recently made changes that reduce the cost of uninitialized replicas. This PR as implemented does also try to push back on senders by throttling an individual sender by 60 seconds after observing one of these rejections. So that places an upper bound on the accumulation of uninitialized replicas at `(num_nodes-1) repls / min`. I'd consider bumping `server.declined_snapshot.timeout` even higher, maybe to 5 minutes.

Between those two factors and the fact that this all only kicks in after read amp 1000 (i.e. the node is already in a very bad state), I think this change will do more good than bad. So I agree that we should proceed with it. 

> Reviewed 3 of 5 files at r1, 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-:-Mt3dYAH8LApozwjNIcH:b-8hslbt)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @erikgrinaker)

___
*[pkg/kv/kvserver/raft.proto, line 210 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-Mt3dC8yEu5tMazIdphR:-Mt3dC8yEu5tMazIdphS:b1hp5rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6382d4a54453fa14211c3b456f0cc1ee6c32f895/pkg/kv/kvserver/raft.proto#L210)):*
> ```protobuf
>     APPLIED = 2;
>     ERROR = 3;
>     DECLINED = 5;
> ```

A lot of this looks like a revert of https://github.com/cockroachdb/cockroach/commit/2a1857f11b93273e2cb4dcf742133000f4e9d0aa. Should we revert that revert and then adjust it with some of the policy changes you've made here? That might help avoid some of the compatibility concerns, because that hasn't made it into a release yet. It also comes with some additional testing which would be valuable here.
___
*[pkg/kv/kvserver/store_snapshot.go, line 610 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-Mt3cncq2DcqFcugUKLh:-Mt3cncq2DcqFcugUKLi:b-s69fa0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6382d4a54453fa14211c3b456f0cc1ee6c32f895/pkg/kv/kvserver/store_snapshot.go#L610)):*
> ```Go
> 
> 	if s.shouldDeclineSnapshot(header) {
> 		return declineSnapshot(stream)
> ```

Do we have any testing that hits this code path?
___
*[pkg/kv/kvserver/store_snapshot.go, line 715 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-Mt3cH1O0lMnPEfgTFsu:-Mt3cH1O0lMnPEfgTFsv:bzcvlio) ([raw file](https://github.com/cockroachdb/cockroach/blob/6382d4a54453fa14211c3b456f0cc1ee6c32f895/pkg/kv/kvserver/store_snapshot.go#L715)):*
> ```Go
> 		&SnapshotResponse{
> 			Status:  SnapshotResponse_DECLINED,
> 			Message: "snapshot declined due to read amplification higher than kv.snapshot_decline.read_amp_threshold",
> ```

There could be other reasons to decline snapshots in the future. For instance, we may decline due to CPU saturation. So we shouldn't be coupling the snapshot status with the reason for the snapshot status. Consider returning this message from `shouldDeclineSnapshot` so we can add other policies later on.
___
*[pkg/kv/kvserver/store_snapshot.go, line 965 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-Mt3ccgU1Bg1Eeo9S5_v:-Mt3ccgU1Bg1Eeo9S5_w:b-t3dmj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6382d4a54453fa14211c3b456f0cc1ee6c32f895/pkg/kv/kvserver/store_snapshot.go#L965)):*
> ```Go
> 			to, snap, resp.Message)
> 	case SnapshotResponse_DECLINED:
> 		// The snapshot was declined by the receiver due to poor LSM health.
> ```

nit: Same point about coupling the status and the reason.


<!-- Sent from Reviewable.io -->

aayushshah15 committed:
Revert "kvserver: remove snapshot CanDecline flag"

This reverts commit 2a1857f11b93273e2cb4dcf742133000f4e9d0aa.

Release note: None

Release justification:

aayushshah15 committed:
kvserver: decline rebalance snapshots on receivers with poor LSM

This commit introduces two new cluster settings:
```
kv.snapshot_decline.read_amp_threshold

server.declined_snapshot.timeout
```

With this commit, stores with a read amplification level higher than
`kv.snapshot_decline.read_amp_threshold` will decline all `REBALANCE`
snapshots. Upon receiving a `DECLINED` response, the senders of these snapshots
will consider these receivers `throttled` for
`server.declined_snapshot.timeout`.

This means that stores with poor LSM health will not be considered as valid
candidates for replica rebalancing.

Fixes #73714
Related to #62168

Release note: None

Release justification: This patch adds a tunable guardrail that could prevent
or mitigate cluster instability

aayushshah15 committed:
kvserver: don't reject raft snapshots on draining nodes

Previously, draining nodes were incorrectly rejecting all snapshots --
including Raft snapshots. This meant that the replicas on those draining nodes
that needed Raft snapshots to catch up would never be able to do so. This
could've lead to tacit unavailability where, even in cases where all the
replicas are live, if a majority is on draining nodes, the range would be
stalled.

Discovered in https://github.com/cockroachlabs/support/issues/1459

Release justification: bug fix

Release note (bug fix): Previously, draining nodes in a cluster without
shutting them down could stall foreground traffic in the cluster. This patch
fixes this bug.

aayushshah15 reviewed (commented):
Sorry for letting this PR sit for so long. It's RFAL now. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-:-MxaB1i4AkDoN7p0R-w3:b-5gfsrb)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/kv/kvserver/store_snapshot.go, line 610 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-Mt3cncq2DcqFcugUKLh:-Mx_YAdB6rf7tWmohpS-:b-qy7sib) ([raw file](https://github.com/cockroachdb/cockroach/blob/6382d4a54453fa14211c3b456f0cc1ee6c32f895/pkg/kv/kvserver/store_snapshot.go#L610)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we have any testing that hits this code path?
</blockquote></details>

Added. 
___
*[pkg/kv/kvserver/store_snapshot.go, line 715 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-Mt3cH1O0lMnPEfgTFsu:-Mx_YDBo376v9eNo-Xh3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6382d4a54453fa14211c3b456f0cc1ee6c32f895/pkg/kv/kvserver/store_snapshot.go#L715)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

There could be other reasons to decline snapshots in the future. For instance, we may decline due to CPU saturation. So we shouldn't be coupling the snapshot status with the reason for the snapshot status. Consider returning this message from `shouldDeclineSnapshot` so we can add other policies later on.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store_snapshot.go, line 965 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-Mt3ccgU1Bg1Eeo9S5_v:-Mx_YERB43QVUWM9X6Nd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6382d4a54453fa14211c3b456f0cc1ee6c32f895/pkg/kv/kvserver/store_snapshot.go#L965)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Same point about coupling the status and the reason.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/raft.proto, line 210 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73720#-Mt3dC8yEu5tMazIdphR:-Mx_YHN97dhTAXlZAZiR:b-uaqyss) ([raw file](https://github.com/cockroachdb/cockroach/blob/6382d4a54453fa14211c3b456f0cc1ee6c32f895/pkg/kv/kvserver/raft.proto#L210)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

A lot of this looks like a revert of https://github.com/cockroachdb/cockroach/commit/2a1857f11b93273e2cb4dcf742133000f4e9d0aa. Should we revert that revert and then adjust it with some of the policy changes you've made here? That might help avoid some of the compatibility concerns, because that hasn't made it into a release yet. It also comes with some additional testing which would be valuable here.
</blockquote></details>

Done. Note that https://github.com/cockroachdb/cockroach/commit/2a1857f11b93273e2cb4dcf742133000f4e9d0aa removed a check to reject snapshots on stores with full disks. That seemed unintentional to me, so I've reintroduced it. 


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
I think we should try to get this and/or #77491 in, since we keep seeing instances of high read amp that could plausibly be explained by unrestrained snapshot ingestion. Since this change differentiates between snapshot types, which was a concern in #77491, maybe we should prioritize this?
